In [1]:
#@title
%%capture
!pip install zipfile36

UsageError: Line magic function `%%capture` not found.


## **Task 2 (20 marks): Sport Analytics** 
Sport analytics has increasingly popular and has changed the way sports are played and player performance is measured. 





### **Dataset description** 

Soccer match event dataset provides to the public the largest open collection of soccer-logs ever released, collected by Wyscout (https://wyscout.com/) containing all the spatio-temporal events (passes, shots, fouls, etc.) that occur during all matches of an entire season of seven competitions (La Liga, Serie A, Bundesliga, Premier League, Ligue 1, FIFA World Cup 2018, UEFA Euro Cup 2016). A match event contains information about its position, time, outcome, player and characteristics.



### **Data collection**
Run the code below to load the datasets to this notebook. 

In [21]:
!wget https://ndownloader.figshare.com/files/14464685/events.zip
import pandas as pd
import zipfile

df_players = pd.read_json("https://ndownloader.figshare.com/files/15073721")

with zipfile.ZipFile("events.zip") as z:
   with z.open("events_Spain.json") as f:
      df_events = pd.read_json(f)

--2022-05-08 01:39:34--  https://ndownloader.figshare.com/files/14464685/events.zip
Resolving ndownloader.figshare.com (ndownloader.figshare.com)... 54.217.124.219, 52.16.102.173, 2a05:d018:1f4:d003:1c8b:1823:acce:812, ...
Connecting to ndownloader.figshare.com (ndownloader.figshare.com)|54.217.124.219|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/14464685/events.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220508/eu-west-1/s3/aws4_request&X-Amz-Date=20220508T013934Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=c17a15682efb767d3998f2e365afce9bdcd90e9020e2e2bd981fc479c7fc348a [following]
--2022-05-08 01:39:34--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/14464685/events.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20220508/eu-west-1/s3/aws4_request&X-Amz-Date=20220508T013934Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X

The code cell above will download two datasets from the archive:


*   Player dataset (`df_players`): describes all players of the teams playing in seven national and international soccer competitions (Italian, Spanish, French, German, English first divisions, World Cup 2018, European Cup 2016). See this [link](https://figshare.com/articles/dataset/Players/7765196?backTo=/collections/Soccer_match_event_dataset/4415000) for details description of each attribute.
*   Even dataset (`df_events`): describes all the events that occur during each match. Each event refers to a ball touch and contains the information such as players involved, event names, and event times. See this [link](https://figshare.com/articles/dataset/Events/7770599?backTo=/collections/Soccer_match_event_dataset/4415000) for details description of each attribute.



In [22]:
# extract data for nationality, full name, and their position in the team
df_players["Nationality"] = [val["name"] for val in df_players["passportArea"]]
df_players["Full Name"] = df_players["firstName"] + " " + df_players["lastName"]
df_players["position"] = [val["name"] for val in df_players["role"]]

df_players.head()

passportArea  weight  \
0  {'name': 'Turkey', 'id': '792', 'alpha3code': ...      78   
1  {'name': 'Senegal', 'id': '686', 'alpha3code':...      73   
2  {'name': 'France', 'id': '250', 'alpha3code': ...      72   
3  {'name': 'Senegal', 'id': '686', 'alpha3code':...      82   
4  {'name': 'France', 'id': '250', 'alpha3code': ...      84   

           firstName middleName     lastName currentTeamId   birthDate  \
0              Harun                   Tekin          4502  1989-06-17   
1             Malang                    Sarr          3775  1999-01-23   
2               Over                Mandanda          3772  1998-10-26   
3  Alfred John Momar                 N'Diaye           683  1990-03-06   
4           Ibrahima             Konat\u00e9          2975  1999-05-25   

   height                                               role  \
0     187  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
1     182  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
2     176  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3     187  {'code2': 'MD', 'code3': 'MID', 'name': 'Midfi...   
4     192  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   

                                           birthArea    wyId   foot  \
0  {'name': 'Turkey', 'id': '792', 'alpha3code': ...   32777  right   
1  {'name': 'France', 'id': '250', 'alpha3code': ...  393228   left   
2  {'name': 'France', 'id': '250', 'alpha3code': ...  393230          
3  {'name': 'France', 'id': '250', 'alpha3code': ...   32793  right   
4  {'name': 'France', 'id': '250', 'alpha3code': ...  393247  right   

        shortName currentNationalTeamId Nationality  \
0        H. Tekin                  4687      Turkey   
1         M. Sarr                  4423     Senegal   
2     O. Mandanda                  null      France   
3      A. N'Diaye                 19314     Senegal   
4  I. Konat\u00e9                  null      France   

                   Full Name    position  
0                Harun Tekin  Goalkeeper  
1                Malang Sarr    Defender  
2              Over Mandanda  Goalkeeper  
3  Alfred John Momar N'Diaye  Midfielder  
4       Ibrahima Konat\u00e9    Defender

In [23]:
# show the first 5 rows of the event table
df_events.head()

eventId subEventName            tags  playerId  \
0        8  Simple pass  [{'id': 1801}]      3542   
1        8  Simple pass  [{'id': 1801}]    274435   
2        8  Simple pass  [{'id': 1801}]    364860   
3        8  Simple pass  [{'id': 1801}]      3534   
4        8  Simple pass  [{'id': 1801}]      3695   

                                  positions  matchId eventName  teamId  \
0  [{'y': 61, 'x': 37}, {'y': 50, 'x': 50}]  2565548      Pass     682   
1  [{'y': 50, 'x': 50}, {'y': 30, 'x': 45}]  2565548      Pass     682   
2  [{'y': 30, 'x': 45}, {'y': 12, 'x': 38}]  2565548      Pass     682   
3  [{'y': 12, 'x': 38}, {'y': 69, 'x': 32}]  2565548      Pass     682   
4  [{'y': 69, 'x': 32}, {'y': 37, 'x': 31}]  2565548      Pass     682   

  matchPeriod   eventSec subEventId         id  
0          1H   2.994582         85  180864419  
1          1H   3.137020         85  180864418  
2          1H   6.709668         85  180864420  
3          1H   8.805497         85  180864421  
4          1H  14.047492         85  180864422

Below in the basic code cell to merge the two datasets for analyses. Columns `playerId` in the `df_events` dataframe and `wyId` in the `df_players` dataframe both are the unique IDs of players in the datasets. The created dataframe `df_all` will be the data used in your tasks.

In [24]:
df_all = df_events.merge(df_players[["wyId", "Full Name", "Nationality", "height", "weight", "position"]], left_on="playerId", right_on="wyId")
df_all.head()

eventId           subEventName                         tags  playerId  \
0        8            Simple pass               [{'id': 1801}]      3542   
1        1  Ground defending duel  [{'id': 701}, {'id': 1802}]      3542   
2        8            Simple pass               [{'id': 1801}]      3542   
3        8              High pass  [{'id': 901}, {'id': 1802}]      3542   
4        8            Simple pass               [{'id': 1801}]      3542   

                                  positions  matchId eventName  teamId  \
0  [{'y': 61, 'x': 37}, {'y': 50, 'x': 50}]  2565548      Pass     682   
1  [{'y': 98, 'x': 89}, {'y': 88, 'x': 57}]  2565548      Duel     682   
2  [{'y': 48, 'x': 39}, {'y': 65, 'x': 26}]  2565548      Pass     682   
3  [{'y': 67, 'x': 32}, {'y': 42, 'x': 74}]  2565548      Pass     682   
4  [{'y': 60, 'x': 35}, {'y': 97, 'x': 42}]  2565548      Pass     682   

  matchPeriod    eventSec subEventId         id  wyId  \
0          1H    2.994582         85  180864419  3542   
1          1H   49.526601         12  180864433  3542   
2          1H  166.859710         85  180864446  3542   
3          1H  181.665639         83  180864450  3542   
4          1H  213.736626         85  180864454  3542   

                     Full Name Nationality  height  weight    position  
0  Manuel Trigueros Mu\u00f1oz       Spain     178      71  Midfielder  
1  Manuel Trigueros Mu\u00f1oz       Spain     178      71  Midfielder  
2  Manuel Trigueros Mu\u00f1oz       Spain     178      71  Midfielder  
3  Manuel Trigueros Mu\u00f1oz       Spain     178      71  Midfielder  
4  Manuel Trigueros Mu\u00f1oz       Spain     178      71  Midfielder

# New Section

The event names are in the "eventName" column.

In [25]:
# show all event names available
df_all["eventName"].unique()

array(['Pass', 'Duel', 'Free Kick', 'Others on the ball', 'Foul', 'Shot',
       'Offside', 'Save attempt', 'Goalkeeper leaving line',
       'Interruption'], dtype=object)

In [26]:
# show all positions available
df_all["position"].unique()

array(['Midfielder', 'Forward', 'Defender', 'Goalkeeper'], dtype=object)

### **Task description**

Create interactive visualisations to analyse this large dataset

2.1 Create visualisation(s) to rank players based on the number of shots they made. Highlight the top 5 players with the most shots. (5 marks)

2.2 Compare the performance (number of shots, number of passes, number of free kicks, number of fouls, number of offsides) of Lionel Messi (wyId: 3359) and Cristiano Ronaldo (wyId: 3322) who are two greatest players in the last decade based on the events in the dataset. (5 marks)

2.3 Create a Datapane report that (1) allows users to rank players based on a selected event (focusing on the 5 events: Pass, Shot, Free Kick, Foul, Offside), and (2) compares the number of passes and the number of shots recorded across all positions ('Midfielder', 'Forward', 'Defender', 'Goalkeeper'). (10 marks)

2.1 Create visualisation(s) to rank players based on the number of shots they made. Highlight the top 5 players with the most shots. (5 marks)

In [27]:
df_all_shot = df_all.loc[df_all['eventName'] == "Shot"]

df_all_shot_groupby = df_all_shot.groupby(["playerId","Full Name","weight","height","Nationality"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()

df_all_shot_groupby["playerId_cat"] = df_all_shot_groupby["playerId"].astype("string")

df_all_shot_groupby=df_all_shot_groupby.sort_values('count', ascending = False)

df_all_shot_groupby

playerId                            Full Name  weight  height  \
20       3322  Cristiano Ronaldo dos Santos Aveiro      83     187   
33       3359  Lionel Andr\u00e9s Messi Cuccittini      72     170   
201      7972   Luis Alberto Su\u00e1rez D\u00edaz      86     182   
178      5400              Gerard Moreno Balaguero      75     180   
109      3840                    Iago Aspas Juncal      67     176   
..        ...                                  ...     ...     ...   
283     70956      R\u00faben Afonso Borges Semedo      84     189   
409    326535               Antonio Latorre Grueso      64     171   
280     70140   Daniel Filipe Martins Carri\u00e7o      80     180   
166      4789         Carles Planas Antol\u00ednez      76     173   
453    545811                    Hugo Duro Perales       0       0   

            Nationality  count playerId_cat  
20             Portugal    151         3322  
33                Spain    142         3359  
201             Uruguay    103         7972  
178               Spain     96         5400  
109               Spain     85         3840  
..                  ...    ...          ...  
283  Cape Verde Islands      1        70956  
409               Spain      1       326535  
280            Portugal      1        70140  
166               Spain      1         4789  
453               Spain      1       545811  

[454 rows x 7 columns]

In [31]:
df_top10 =  df_all_shot_groupby.head(10)


In [28]:
df_top5 =  df_all_shot_groupby.head(5)
df_top5

playerId                            Full Name  weight  height  \
20       3322  Cristiano Ronaldo dos Santos Aveiro      83     187   
33       3359  Lionel Andr\u00e9s Messi Cuccittini      72     170   
201      7972   Luis Alberto Su\u00e1rez D\u00edaz      86     182   
178      5400              Gerard Moreno Balaguero      75     180   
109      3840                    Iago Aspas Juncal      67     176   

    Nationality  count playerId_cat  
20     Portugal    151         3322  
33        Spain    142         3359  
201     Uruguay    103         7972  
178       Spain     96         5400  
109       Spain     85         3840

In [34]:
top5list=df_top5['Full Name'].tolist()
print(top5list)

['Cristiano Ronaldo dos Santos Aveiro', 'Lionel Andr\\u00e9s Messi Cuccittini', 'Luis Alberto Su\\u00e1rez D\\u00edaz', 'Gerard Moreno Balaguero', 'Iago Aspas Juncal']


In [36]:
fig = px.bar(df_top10, x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Shot"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig.update_layout(
    title={
        'text': "Ranking of players based on number of shots",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig["data"][0]["marker"]["color"] = ["red" if playerId_cat in top5list else "blue" for playerId_cat in fig["data"][0]["x"]]
fig

1. This visualisation is ranking the players based on number of shots using barchart. Red colouring has been used to highlight the top 5 players with most shots out of the top 10 players. Hovering over the individual bars will reveal the player's details such as height, weight, nationality and shot.

2.2 Compare the performance (number of shots, number of passes, number of free kicks, number of fouls, number of offsides) of Lionel Messi (wyId: 3359) and Cristiano Ronaldo (wyId: 3322) who are two greatest players in the last decade based on the events in the dataset. (5 marks)

In [37]:
list_top2 = (3359,3322)

In [38]:
eventName_list = ("Pass", "Foul", "Shot", "Free Kick", "Offside")

In [39]:
df_all_top2 = df_all.loc[df_all['wyId'].isin(list_top2)]
df_all_top2 = df_all_top2.loc[df_all_top2['eventName'].isin(eventName_list)]
df_all_top2

eventId subEventName  \
232426        8  Simple pass   
232427        8  Simple pass   
232430        8  Simple pass   
232432       10         Shot   
232434        8  Simple pass   
...         ...          ...   
528860        8    Head pass   
528862        8  Simple pass   
528864        8  Simple pass   
528865        8  Simple pass   
528866        8  Simple pass   

                                                     tags  playerId  \
232426                                     [{'id': 1801}]      3359   
232427                                     [{'id': 1801}]      3359   
232430                                     [{'id': 1801}]      3359   
232432  [{'id': 401}, {'id': 201}, {'id': 1215}, {'id'...      3359   
232434                                     [{'id': 1801}]      3359   
...                                                   ...       ...   
528860                                     [{'id': 1801}]      3322   
528862                                     [{'id': 1801}]      3322   
528864                                     [{'id': 1801}]      3322   
528865                                     [{'id': 1801}]      3322   
528866                                     [{'id': 1802}]      3322   

                                       positions  matchId eventName  teamId  \
232426  [{'y': 50, 'x': 50}, {'y': 50, 'x': 40}]  2565554      Pass     676   
232427  [{'y': 64, 'x': 71}, {'y': 67, 'x': 54}]  2565554      Pass     676   
232430  [{'y': 74, 'x': 83}, {'y': 61, 'x': 77}]  2565554      Pass     676   
232432    [{'y': 61, 'x': 82}, {'y': 0, 'x': 0}]  2565554      Shot     676   
232434  [{'y': 27, 'x': 70}, {'y': 33, 'x': 71}]  2565554      Pass     676   
...                                          ...      ...       ...     ...   
528860  [{'y': 50, 'x': 31}, {'y': 69, 'x': 23}]  2565927      Pass     675   
528862  [{'y': 77, 'x': 55}, {'y': 76, 'x': 40}]  2565927      Pass     675   
528864  [{'y': 54, 'x': 56}, {'y': 95, 'x': 68}]  2565927      Pass     675   
528865  [{'y': 84, 'x': 53}, {'y': 96, 'x': 63}]  2565927      Pass     675   
528866  [{'y': 89, 'x': 73}, {'y': 83, 'x': 71}]  2565927      Pass     675   

       matchPeriod    eventSec subEventId         id  wyId  \
232426          1H    1.012047         85  180465950  3359   
232427          1H   51.068905         85  180465968  3359   
232430          1H  204.912191         85  180466018  3359   
232432          1H  217.537011        100  180466023  3359   
232434          1H  362.197211         85  180466052  3359   
...            ...         ...        ...        ...   ...   
528860          2H   74.391650         82  253301599  3322   
528862          2H  125.939463         85  253301622  3322   
528864          2H  540.156764         85  253301797  3322   
528865          2H  737.846860         85  253301860  3322   
528866          2H  741.389212         85  253301865  3322   

                                  Full Name Nationality  height  weight  \
232426  Lionel Andr\u00e9s Messi Cuccittini       Spain     170      72   
232427  Lionel Andr\u00e9s Messi Cuccittini       Spain     170      72   
232430  Lionel Andr\u00e9s Messi Cuccittini       Spain     170      72   
232432  Lionel Andr\u00e9s Messi Cuccittini       Spain     170      72   
232434  Lionel Andr\u00e9s Messi Cuccittini       Spain     170      72   
...                                     ...         ...     ...     ...   
528860  Cristiano Ronaldo dos Santos Aveiro    Portugal     187      83   
528862  Cristiano Ronaldo dos Santos Aveiro    Portugal     187      83   
528864  Cristiano Ronaldo dos Santos Aveiro    Portugal     187      83   
528865  Cristiano Ronaldo dos Santos Aveiro    Portugal     187      83   
528866  Cristiano Ronaldo dos Santos Aveiro    Portugal     187      83   

       position  
232426  Forward  
232427  Forward  
232430  Forward  
232432  Forward  
232434  Forward  
...         ...  
528860  Forward  
528862  Forward  
528864  Forwa

In [40]:
df_all_top2["wyId"] = df_all_top2["wyId"].astype("category")
df_all_top2_groupby = df_all_top2.groupby(["wyId","eventName","Full Name"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()
df_all_top2_groupby

wyId  eventName                            Full Name  count
0   3322       Foul  Cristiano Ronaldo dos Santos Aveiro     18
1   3322       Foul  Lionel Andr\u00e9s Messi Cuccittini      0
2   3322  Free Kick  Cristiano Ronaldo dos Santos Aveiro     42
3   3322  Free Kick  Lionel Andr\u00e9s Messi Cuccittini      0
4   3322    Offside  Cristiano Ronaldo dos Santos Aveiro     39
5   3322    Offside  Lionel Andr\u00e9s Messi Cuccittini      0
6   3322       Pass  Cristiano Ronaldo dos Santos Aveiro    727
7   3322       Pass  Lionel Andr\u00e9s Messi Cuccittini      0
8   3322       Shot  Cristiano Ronaldo dos Santos Aveiro    151
9   3322       Shot  Lionel Andr\u00e9s Messi Cuccittini      0
10  3359       Foul  Cristiano Ronaldo dos Santos Aveiro      0
11  3359       Foul  Lionel Andr\u00e9s Messi Cuccittini     18
12  3359  Free Kick  Cristiano Ronaldo dos Santos Aveiro      0
13  3359  Free Kick  Lionel Andr\u00e9s Messi Cuccittini    179
14  3359    Offside  Cristiano Ronaldo dos Santos Aveiro      0
15  3359    Offside  Lionel Andr\u00e9s Messi Cuccittini     13
16  3359       Pass  Cristiano Ronaldo dos Santos Aveiro      0
17  3359       Pass  Lionel Andr\u00e9s Messi Cuccittini   1787
18  3359       Shot  Cristiano Ronaldo dos Santos Aveiro      0
19  3359       Shot  Lionel Andr\u00e9s Messi Cuccittini    142

In [41]:
fig = px.bar(df_all_top2_groupby, x="eventName", y = "count", color="Full Name", barmode="group")
fig.update_layout(
    title={
        'text': "Compare the performance of Lionel Messi and Cristiano Ronaldo",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig

1.   This visualisation is comparing the performance (number of shots, number of passes, number of free kicks, number of fouls, number of offsides) of Lionel Messi (wyId: 3359) and Cristiano Ronaldo (wyId: 3322) using barchart. Colouring has been used to discriminate between the 2 players.
2.   According to the graph, Lionel performed simiarly to Cristiano at 18 fouls. In terms of free kicks, Lionel had more then 4 times higher free kicks compared to Cristiano. In terms of offsides, Lionel performed 3 times better than Cristiano. In terms of pases, Lionel performed about 2.4 times better than Cristiano. However, Cristiano had higher shot at 151 compared to Lionel at 142.



2.3 Create a Datapane report that (1) allows users to rank players based on a selected event (focusing on the 5 events: Pass, Shot, Free Kick, Foul, Offside), and (2) compares the number of passes and the number of shots recorded across all positions ('Midfielder', 'Forward', 'Defender', 'Goalkeeper'). (10 marks)

In [42]:
# %%capture
!pip install datapane
!pip install pyyaml==5.4.1
!datapane login --token=8bb52c70c01673477f8726d60577c863f072f14e
import plotly.express as px
import pandas as pd
import datapane as dp

     |████████████████████████████████| 2.6 MB 5.3 MB/s 
     |████████████████████████████████| 97 kB 6.9 MB/s 
     |████████████████████████████████| 46 kB 4.1 MB/s 
     |████████████████████████████████| 45 kB 3.7 MB/s 
     |████████████████████████████████| 54 kB 2.8 MB/s 
     |████████████████████████████████| 193 kB 48.5 MB/s 
     |████████████████████████████████| 596 kB 40.0 MB/s 
     |████████████████████████████████| 10.9 MB 54.6 MB/s 
     |████████████████████████████████| 46 kB 3.7 MB/s 
  Created wheel for face: filename=face-20.1.1-py3-none-any.whl size=51091 sha256=ca69ac1a9f09c1558007d149411e9d6bf4e3aa89b68fba29bb183d50beefa505
  Stored in directory: /root/.cache/pip/wheels/67/03/c2/76d2e1e694852ce409ac2938b9f461bc0fdf3789901e7bb1f0
  Created wheel for micawber: filename=micawber-0.5.4-py3-none-any.whl size=22373 sha256=494a457d3e7a61402a5d0d041f75df6496bd32652b87c0fc0a1c535a829fbd17
  Stored in directory: /root/.cache/pip/wheels/c1/b6/81/f8237c6061ac239fe1079fa3

     |████████████████████████████████| 636 kB 5.0 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 6.0
    Uninstalling PyYAML-6.0:
      Successfully uninstalled PyYAML-6.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datapane 0.14.0 requires PyYAML<7.0.0,>=6.0.0, but you have pyyaml 5.4.1 which is incompatible.


Connected successfully to https://datapane.com as milo.joronen@gmail.com


In [43]:
df_all_pass = df_all.loc[df_all['eventName'] == "Pass"]

df_all_pass_groupby = df_all_pass.groupby(["playerId","Full Name","weight","height","Nationality"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()

df_all_pass_groupby["playerId_cat"] = df_all_pass_groupby["playerId"].astype("string")

df_all_pass_groupby=df_all_pass_groupby.sort_values('count', ascending = False)

df_all_pass_groupby

playerId                          Full Name  weight  height Nationality  \
70       3476                  Ivan Rakiti\u0107      78     184     Croatia   
102      3676        Asier Illarramendi Andonegi      75     179       Spain   
507    364860    Rodrigo Hern\u00e1ndez Cascante      78     190       Spain   
12       3286           Daniel Parejo Mu\u00f1oz      74     182       Spain   
7        3269                   Jordi Alba Ramos      68     170       Spain   
..        ...                                ...     ...     ...         ...   
231      6925  H\u00e9ctor Hern\u00e1ndez Ortega      71     173       Spain   
424    257752           Marlos Moreno Dur\u00e1n      67     173    Colombia   
530    400564           Dar\u00edo Poveda Romera      75     185       Spain   
484    329026    Carlos Calder\u00f3n L\u00f3pez      69     182       Spain   
510    367391                  Antonio Moya Vega       0       0       Spain   

     count playerId_cat  
70    2417         3476  
102   2396         3676  
507   2338       364860  
12    2153         3286  
7     2089         3269  
..     ...          ...  
231      1         6925  
424      1       257752  
530      1       400564  
484      1       329026  
510      1       367391  

[556 rows x 7 columns]

In [44]:
df_all_freekick = df_all.loc[df_all['eventName'] == "Free Kick"]

df_all_freekick_groupby = df_all_freekick.groupby(["playerId","Full Name","weight","height","Nationality"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()

df_all_freekick_groupby["playerId_cat"] = df_all_freekick_groupby["playerId"].astype("string")

df_all_freekick_groupby=df_all_freekick_groupby.sort_values('count', ascending = False)

df_all_freekick_groupby

playerId                                         Full Name  weight  \
148      3974  Dami\u00e1n Nicol\u00e1s Su\u00e1rez Su\u00e1rez      61   
18       3299                           Fernando Pacheco Flores      81   
3         254                  Roberto Jos\u00e9 Rosales Altuve      71   
326     93000                 Vitorino Gabriel  Pacheco Antunes      70   
114      3719                         Roberto Jim\u00e9nez Gago      93   
..        ...                                               ...     ...   
185      4440                               Borja Valle Balonga      73   
438    293857                           Alexander Mesa Travieso      77   
145      3937                         Roger Mart\u00ed Salvador      74   
445    334817                             Lucas Ariel Boy\u00e9      80   
492    516397                                Mujaid Sadick Aliu       0   

     height Nationality  count playerId_cat  
148     172     Uruguay    534         3974  
18      186       Spain    466         3299  
3       175   Venezuela    431          254  
326     176    Portugal    428        93000  
114     192       Spain    425         3719  
..      ...         ...    ...          ...  
185     177       Spain      1         4440  
438     178       Spain      1       293857  
145     179       Spain      1         3937  
445     180   Argentina      1       334817  
492     180       Spain      1       516397  

[493 rows x 7 columns]

In [45]:
df_all_foul = df_all.loc[df_all['eventName'] == "Foul"]

df_all_foul_groupby = df_all_foul.groupby(["playerId","Full Name","weight","height","Nationality"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()

df_all_foul_groupby["playerId_cat"] = df_all_foul_groupby["playerId"].astype("string")

df_all_foul_groupby=df_all_foul_groupby.sort_values('count', ascending = False)

df_all_foul_groupby

playerId                               Full Name  weight  height  \
107      3714        Cristhian Ricardo Stuani Curbelo      72     184   
68       3503          Ra\u00fal Garc\u00eda Escudero      83     184   
471    364860         Rodrigo Hern\u00e1ndez Cascante      78     190   
461    344140                   Amath Diedhiou Ndiaye      63     173   
90       3650                    Tom\u00e1s Pina Isla      75     185   
..        ...                                     ...     ...     ...   
433    282445                   Ramiro Guerra Pereyra      75     183   
445    298385             Borja Herrera Gonz\u00e1lez      70     178   
453    339857  Jos\u00e9 Manuel Arn\u00e1iz D\u00edaz      70     175   
464    352748       Iv\u00e1n Rodr\u00edguez del Pozo      70     180   
426    279560               Andoni Gorosabel Espinosa      73     174   

    Nationality  count playerId_cat  
107       Italy     87         3714  
68        Spain     81         3503  
471       Spain     79       364860  
461     Senegal     74       344140  
90        Spain     72         3650  
..          ...    ...          ...  
433     Uruguay      1       282445  
445       Spain      1       298385  
453       Spain      1       339857  
464       Spain      1       352748  
426       Spain      1       279560  

[511 rows x 7 columns]

In [46]:
df_all_offside = df_all.loc[df_all['eventName'] == "Offside"]

df_all_offside_groupby = df_all_offside.groupby(["playerId","Full Name","weight","height","Nationality"]).agg(
    count=pd.NamedAgg(column="eventName", aggfunc="count")
).reset_index()

df_all_offside_groupby["playerId_cat"] = df_all_offside_groupby["playerId"].astype("string")

df_all_offside_groupby=df_all_offside_groupby.sort_values('count', ascending = False)

df_all_offside_groupby

playerId                                     Full Name  weight  height  \
147      7972            Luis Alberto Su\u00e1rez D\u00edaz      86     182   
272    278278                      Munir El Haddadi Mohamed      69     175   
16       3322           Cristiano Ronaldo dos Santos Aveiro      83     187   
239    228902                              Jonathan Calleri      70     173   
213    105541                Lucas P\u00e9rez Mart\u00ednez      73     180   
..        ...                                           ...     ...     ...   
42       3470  Jos\u00e9 \u00c1ngel G\u00f3mez Campa\u00f1a      70     179   
77       3735             Juan Carlos P\u00e9rez L\u00f3pez      70     179   
46       3510                       David Timor Copov\u00ed      80     186   
233    225022                Rub\u00e9n Duarte S\u00e1nchez      75     180   
109      4329                     Alejandro Arribas Garrido      76     182   

    Nationality  count playerId_cat  
147     Uruguay     55         7972  
272     Morocco     45       278278  
16     Portugal     39         3322  
239       Italy     38       228902  
213       Spain     38       105541  
..          ...    ...          ...  
42        Spain      1         3470  
77        Spain      1         3735  
46        Spain      1         3510  
233       Spain      1       225022  
109       Spain      1         4329  

[314 rows x 7 columns]

In [49]:
df_shot_top10 =  df_all_shot_groupby.head(10)

In [75]:
fig_shot = px.bar(df_shot_top10, x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Shot"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig_shot.update_layout(
    title={
        'text': "Ranking of top 10 players based on number of shots",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})

This visualisation is ranking the players based on number of shots using barchart for the top 10 players in descending order.

In [55]:
df_pass_top10 =  df_all_pass_groupby.head(10)

In [76]:
fig_pass = px.bar(df_pass_top10, x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Pass"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig_pass.update_layout(
    title={
        'text': "Ranking of top 10 players based on number of passes",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})

This visualisation is ranking the players based on number of passes using barchart for the top 10 players in descending order.

In [57]:
df_freekick_top10 =  df_all_freekick_groupby.head(10)

In [77]:
fig_freekick = px.bar(df_freekick_top10, x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Free Kick"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig_freekick.update_layout(
    title={
        'text': "Ranking of top 10 players based on number of free kicks",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})

This visualisation is ranking the players based on number of free kicks using barchart for the top 10 players in descending order.

In [60]:
df_foul_top10 =  df_all_foul_groupby.head(10)

In [78]:
fig_foul = px.bar(df_foul_top10, x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Foul"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig_foul.update_layout(
    title={
        'text': "Ranking of top 10 players based on number of fouls",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})

This visualisation is ranking the players based on number of fouls using barchart for the top 10 players in descending order.

In [63]:
df_offside_top10 =  df_all_offside_groupby.head(10)

In [79]:
fig_offside = px.bar(df_offside_top10 , x="Full Name", y="count",
                labels={
                     "playerId_cat": "Player ID",
                     "count": "Offside"
                     
                 }, hover_name= "Full Name", hover_data=["weight","height","Nationality"])
fig_offside.update_layout(
    title={
        'text': "Ranking of top 10 players based on number of offsides",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})

This visualisation is ranking the players based on number of offsides using barchart for the top 10 players in descending order.

In [80]:
report = dp.Report(
    "## rank players based on a selected event(focusing on the 5 events: Pass, Shot, Free Kick, Foul, Offside)",
    dp.Select(
            dp.Plot(fig_shot, label="Ranking by shots"),
            dp.Plot(fig_pass, label="Ranking by passes"),
            dp.Plot(fig_freekick, label="Ranking by free kicks"),
            dp.Plot(fig_foul, label="Ranking by fouls"),
            dp.Plot(fig_offside, label="Ranking by offsides"),
            type=dp.SelectType.TABS
    ),
)
report.upload(name="rank players based on a selected event")


Uploading report and associated data - *please wait...*
Your report only contains a single element - did you know you can include additional plots, tables and text in a single report? More info at https://docs.datapane.com/reports/blocks/layout-pages-and-selects
Report successfully uploaded. View and share your report at https://datapane.com/reports/M3820lk/rank-players-based-on-a-selected-event/, or edit your report at https://datapane.com/reports/M3820lk/rank-players-based-on-a-selected-event/edit/.


compares the number of
passes and the number of shots recorded across all positions ('Midfielder', 'Forward', 'Defender',
'Goalkeeper').

In [67]:
df_all_position = df_all.groupby(["position"]).agg(
    count=pd.NamedAgg(column="position", aggfunc="count")
).reset_index()
df_all_position

position   count
0    Defender  226338
1     Forward   97364
2  Goalkeeper   28754
3  Midfielder  229522

In [68]:
df_all_pass = df_all.loc[df_all['eventName'] == "Pass"]

df_all_pass_position_groupby = df_all_pass.groupby(["position","eventName"]).agg(
    count=pd.NamedAgg(column="position", aggfunc="count")
).reset_index()

df_all_pass_position_groupby["Percentage"] = round((df_all_pass_position_groupby["count"] / df_all_position.groupby(["position"])["count"].transform('sum') *100),2)

df_all_pass_position_groupby

position eventName   count  Percentage
0    Defender      Pass  131317       58.02
1     Forward      Pass   40644       41.74
2  Goalkeeper      Pass   14010       48.72
3  Midfielder      Pass  132742       57.83

In [69]:
fig_pass_position = px.bar(df_all_pass_position_groupby, x="position", y = "Percentage", color="position")
fig_pass_position.update_layout(
    title={
        'text': "Compare passes across all positions",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig_pass_position

1.   This visualisation is comparing passes (in percentages) using barchart. Colouring has been used to discriminate between the all the positions.
2.   According to the graph, on average, a defender makes the most passes at 58.02%, followed by a midfiedler at 57.83%. A goal keeper comes third at 48.72% and followed by a forward at 41.74%.

In [70]:
df_all_shot = df_all.loc[df_all['eventName'] == "Shot"]

df_all_shot_position_groupby = df_all_shot.groupby(["position","eventName"]).agg(
    count=pd.NamedAgg(column="position", aggfunc="count")
).reset_index()

df_all_shot_position_groupby["Percentage"] = round((df_all_shot_position_groupby["count"] / df_all_position.groupby(["position"])["count"].transform('sum') *100),2)

df_all_shot_position_groupby

position eventName  count  Percentage
0    Defender      Shot   1263        0.56
1     Forward      Shot   3475        3.57
2  Goalkeeper      Shot      2        0.01
3  Midfielder      Shot   3238        1.41

In [71]:
fig_shot_position = px.bar(df_all_shot_position_groupby, x="position", y = "Percentage", color="position")
fig_shot_position.update_layout(
    title={
        'text': "Compare shots across all positions",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
fig_shot_position

1.   This visualisation is comparing shots (in percentages) using barchart. Colouring has been used to discriminate between the all the positions.
2.   According to the graph, on average, a forward makes the most shots at 3.57%, followed by a midfiedler at 1.41%. A defender comes third at 0.56% and followed by a goal keeper at 0.01%.